### ISSUE TITLE
OpenID token too large with roles inside

### ISSUE THREAD

**Issue by cedef** [mrparkers/terraform-provider-keycloak#483]
Hi,

First of all, thanks your for work on this provider !

I've started to setup a keycloak instance behind a nginx reverse proxy, with ~30 realms.
I've setup the terraform client using OpenID + serviceAccount role "admin", and I'm getting a `400 Request Header Or Cookie Too Large` error.

When I reproduce this setup locally (i.e. without nginx proxy), I can see big access token, listing all roles for every realm.... such token can be quite large (~10-30KiB) and worst: it will grow accordingly with the number of realms (!).

Increasing nginx `proxy_buffer*` parameters is not really a solutionto me as I'm going to manage douzens of realms with terraform... It means that request header could be as large as 60~128KiB or even more...

On the keycloak side removing roles from "client scopes" leads terraform provider to a `403 Forbidden` error.

Is there a solution to reduce the token size while still having this provider working ?
Is there a way to query client roles in a separate API query if it's not included in token ?

Thank you,

**Comment by dmeyerholt** [comment]
Hi,
Try adjusting the role scope mappings of that client. By default they are set to full scope meaning every single role of that Service Account will be put in the issued access tokens of that Client.

**Comment by tomrutsaert** [comment]
If I recall correctly, You should use the admin-cli client.
The token you get for the admin-cli client will be smaller, because it is missing the resource_access and realm_access part.
But it will still be able to do the call
The keycloak admin rest api code has a different logic path when the incoming token is for the admin-c;i client, it does not need the resource_access and realm_access claims. It checks the acces/roles in a different way. I can not recall how exactly.

**Comment by cedef** [comment]
@dmeyerholt :
> Try adjusting the role scope mappings of that client. By default they are set to full scope meaning every single role of that Service Account will be put in the issued access tokens of that Client. However roles are checked by the admin rest endpoints and they rely on roles to be present in the token.

That's what I've attempted to do without success:

- with only "admin" as "Assigned Roles" (see image)![image](https://user-images.githubusercontent.com/1536135/109519345-6f494080-7aab-11eb-9a90-0c23240be753.png)  I'm still getting a very long access token and in the end:
 ```error sending GET request to /auth/admin/serverinfo: 400 Bad Request. Response
body: <html><head><title>400 Request Header Or Cookie Too Large</title></head>
```

- with nothing in Assigned Roles 
![image](https://user-images.githubusercontent.com/1536135/109519696-c222f800-7aab-11eb-880e-ac8bbcba00f8.png), I'm getting a 403 Forbidden. In that case, the decoded JWT token contains:
```JWT Header:
{
  "alg": "RS256",
  "typ": "JWT",
  "kid": "~ truncated ~"
}
JWT Payload:
{
  "exp": 1614613855,
  "iat": 1614613795,
  "jti": "~ truncated ~",
  "iss": "https://keycloak.localhost/auth/realms/master",
  "sub": "~ truncated ~",
  "typ": "Bearer",
  "azp": "terraform",
  "acr": "1",
  "scope": "email profile",
  "clientId": "terraform",
  "email_verified": false,
  "clientHost": "::1",
  "preferred_username": "service-account-terraform",
  "clientAddress": "::1"
}
```

- If I give all roles only for Master-realm 
![image](https://user-images.githubusercontent.com/1536135/109522651-c866a380-7aae-11eb-9c8e-288cb5d50559.png), I still get a forbidden and the following token:
```
JWT Header:                                                                                                                           [4/1812]
{                                                                                                                                             
  "alg": "RS256",                                          
  "typ": "JWT",                                                      
  "kid": "~ truncated ~"                                                                                        
}
JWT Payload:
{
  "exp": 1614614040,
  "iat": 1614613980,
  "jti": "~ truncated ~",
  "iss": "https://keycloak.localhost/auth/realms/master",
  "aud": "master-realm",
  "sub": "~ truncated ~",
  "typ": "Bearer",
  "azp": "terraform",
  "acr": "1",
  "resource_access": {
    "master-realm": {
      "roles": [
        "view-realm",
        "view-identity-providers",
        "manage-identity-providers",
        "impersonation",
        "create-client",
        "manage-users",
        "query-realms",
        "view-authorization",
        "query-clients",
        "query-users",
        "manage-events",
        "manage-realm",
        "view-events",
        "view-users",
        "view-clients",
        "manage-authorization",
        "manage-clients",
        "query-groups"
      ]
    }
  },
  "scope": "email profile",
  "clientId": "terraform",
  "email_verified": false,
  "clientHost": "::1",
  "preferred_username": "service-account-terraform",
  "clientAddress": "::1"
}
```

Maybe am I doing something wrong ?

On the other side, using `admin-cli` (switching to confidential + enabling service Account + adding "admin" into service account roles) do the job...

Thank you for your help !

**Comment by dmeyerholt** [comment]
Hi @cedef  ,
you are doing nothing wrong. actually we also ran into a similar problem today (not tf provider related) when using such "admin-role" bearer token through an ingress. It was only then we realized how big our own token is (~75k). One problem here is that the "admin" realm role is a composite role including all realm-clients' client roles. When creating a new realm, this composite role automatically gets enhanced with the new roles of the new "realm-client". 
Then the "roles" Client scope maps all those client roles into the access token (and audiences as well). It may be possible to configure an own roles mapper which only respects certain roles but this seems very tedious and is imho unmanageable.

Like @tomrutsaert mentioned the keycloak-code has hard-coded handling for the admin-cli. I can be seen [here](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/resources/admin/permissions/MgmtPermissions.java#L99) for example.

We are currently investigating possible mitigations but i think in the end your solution (switch admin-cli to confidential and enable service-account) is the only way to go here.

Will come back here if we found another solution.

Best
